Source: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/3105
Author: two <two@envs.net>
Editor: Mingi Sung <fiestalake@disroot.org>
Commit: ee3783d0b62f68b7e151a302127a6bbbd73e72f8
Last Updated: 8/2/23 (Mutter 44.3)
Date: Sun, 2 Jul 2023 12:46:17 +0000
Subject: [PATCH] core: Only avoid focusing x11 windows on map
---

in issue #2690 people say the bug that was fixed by adding this check was only on x11, while it introduced another bug, focus is sometimes not given when launching a window from overview.
if it's only checked on X11, it can be at least avoided on wayland.

---
 src/core/window.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)
 
diff --git a/src/core/window.c b/src/core/window.c
index 6c9247743..324db7e30 100644
--- a/src/core/window.c
+++ b/src/core/window.c
@@ -2007,10 +2007,10 @@ window_state_on_map (MetaWindow *window,
       return;
     }
 
-  /* Do not focus window on map if input is already taken by the
-   * compositor.
+  /* Do not focus x11 window on map if input is already taken by the
+   * compositor (wayland windows can't grab input anyway).
    */
-  if (!meta_display_windows_are_interactable (window->display))
+  if (!meta_display_windows_are_interactable && window->client_type == META_WINDOW_CLIENT_TYPE_X11)
     {
       *takes_focus = FALSE;
       return;
